TST use global_random_seed in sklearn/cluster/tests/test_hierarchical.py

marenwestermann opened the pull request:
#### Reference Issues/PRs
towards #22827


#### What does this implement/fix? Explain your changes.
I used the global_random_seed fixture in the `sklearn/cluster/tests/test_hierarchical.py` module where I thought it's appropriate in accordance with the description of the referenced issue. 

A few comments about tests where I at first thought the global_random_seed fixture would be useful but didn't apply it:
 
`test_agglomerative_clustering_memory_mapped` and `test_mst_linkage_core_memory_mapped`.
In the documentation of the numpy `memmap` function on which these tests are based it is said, "Memory-mapped files are used for accessing small segments of large files on disk, without reading the entire file into memory." (for reference see [here](https://numpy.org/doc/stable/reference/generated/numpy.memmap.html)). Therefore I thought the global_random_seed fixture could be useful in these two tests because changing the seed might result in a different small segment being chosen (but I'm not sure if this is correct). When I applied this fixture to these two tests, all tests passed but the former test ran in 1.83 s on my machine (Apple M1 Pro) and the latter ran in 14 s. Because of the rather long run time, I didn't apply the fixture in these tests.

I also tried applying the fixture to the function `test_ward_linkage_tree_return_distance` and all tests ran successfully. However, the run time on my machine was 1.4 s and I therefore decided against applying this fixture here. 

#### Any other comments?
#PyLadiesParis
I started working on this issue during the PyLadies Paris scikit-learn sprint on 8 Dec 2022.

marenwestermann committed:
test_hierarchical [all random seeds]
test_linkage_misc
test_ward_agglomeration
test_ward_tree_children_order
test_small_distance_threshold
test_cluster_distances_with_distance_threshold

glemaitre reviewed (commented):
I would add the fixture to the following tests.

- `test_agglomerative_clustering`
- `test_sparse_scikit_vs_scipy`
- `test_vector_scikit_single_vs_scipy_single` this on is already parameterized with some seed. We can remove them.
- `test_vector_scikit_single_vs_scipy_single`
- `test_agglomerative_clustering_with_distance_threshold`

Somehow these tests check some equivalence and I think this is worth checking with different inputs to find if there are some corner cases.

glemaitre commented:
I think this is fine to let the test with the memmap out. For the other test since we are running by default a single seed this is fine to have an increase time because it would only be triggered if we enable the test for it.

marenwestermann committed:
test-hierarchical [all random seeds]
test_sparse_scikit_vs_scipy
test_vector_scikit_single_vs_scipy_single
test_ward_linkage_tree_return_distance
test_agglomerative_clustering_with_distance_threshold

marenwestermann commented:
Thanks for your review @glemaitre. I updated the PR. When I added the global_random_seed fixture to the test `test_agglomerative_clustering` the run time increased from previously 0.4 sec to 12 sec, so I didn't apply the fixture for this test. If you think that's ok I'll add the fixture to this test as well.

glemaitre commented:
We can add the fixture to this test. The increased time only happens when we request it for this test specifically which is never the case. Here, we only make sure that the test is passing locally whatever the seed. So this is fine.

marenwestermann committed:
test-hierarchical [all random seeds]
test_agglomerative_clustering

glemaitre reviewed (approved):
LGTM. Thanks @marenwestermann.
Noting as awaiting a second review.

thomasjpfan reviewed (approved):
Thank you for the PR! LGTM
